### PR TITLE
Add the first unit-test-passed version.

### LINKED ISSUE #2: Merge a preliminary version into develop
This version should support building index and searching. But it should not support the acceleration of search using $UB_t$ as explained in the [WEAK-AND paper](http://cis.poly.edu/westlab/papers/cntdstrb/p426-broder.pdf). So that after this issue is solved, we can have multiple branches out from develop to test various bounding/acceleration strategies.

### LINKED ISSUE #3: Test correctness on Chinese data
Currently, we have in [`dev` branch](https://github.com/wangkuiyi/weakand/tree/dev) we have English test data in `testdata/aaai14papers.txt`. But we also need to test using Chinese data.

### LINKED ISSUE #4: Write a frontend (demo) server and a backend server
Currently on [`dev` branch](https://github.com/wangkuiyi/weakand/tree/dev), we develop a Go package implementing indexing and searching functionalities. Then we need to write a serving system that use this package.

The serving system should includes two programs: 
1. a frontend (HTTP) server which provides two Web pages:
   1. a page allowing users to input a new document. After clicking the Submit button, the Web browser sends this text to the frontend server, and frontend proxy this text via Go RPC to the backend server, which maintains and updates the search index data structure.
   2. a search page allowing users to input a query. After clicking the Submit button, the frontend server should display search results.
2. a backend (Go RPC) server which provides two RPC calls that will be called by the frontend server:
   1. `Index.Add`: add a new document into the index data structure.
   2. `Index.Search`: accepts a query and returns ranked search results.

When the system is deployed, it would be very likely that clients' frontend server will call our backend server. Our frontend server is only for demo.

### PR DESCRIPTION
**Pull request by wangkuiyi**
This change adds the first unit-test passed version.  Your comments are highly appreciated!

This change fixes https://github.com/wangkuiyi/weakand/issues/2 and fixes https://github.com/wangkuiyi/weakand/issues/3 and fixes https://github.com/wangkuiyi/weakand/issues/4

You might want to checkout the code and run tests:

```
go get github.com/wangkuiyi/weakand
cd $GOPATH/src/github.com/huichen/sego
git remote add wangkuiyi https://github.com/wangkuiyi/sego.git
git fetch wangkuiyi
git checkout fs
go get github.com/wangkuiyi/fs
cd $GOPATH/src/github.com/wangkuiyi/weakand
git checkout dev
go install && go test -v && go vet
```

Please be aware that weakand uses github.com/huichen/sego, a Chinese segmenter, but needs some changes not yet merged into its master branch. So, after `go get github.com/wangkuiyi/weakand`, we might need to switch to a feature branch located at `https://github.com/wangkuiyi/sego:/ref/fs`.

_Update:_

I wrote a frontend server and a backend server as wanted by https://github.com/wangkuiyi/weakand/issues/4

To build and run these server:

```
go install github.com/wangkuiyi/weakand/frontend
go install github.com/wangkuiyi/weakand/backend
$GOPATH/bin/backend -corpus=$GOPATH/src/github.com/wangkuiyi/weakand/testdata/aaai14papers.txt &
$GOPATH/bin/frontend &
```

then you will be able to open a Web browser tab and enter the URL `localhost:18081` to search AAAI 2014 paper titles.

You can also visit `http://localhost:18081/add` to type a new paper title there. For example: 哎哟我去. This new document will be inserted into the in-memory index. So if you visit `localhost:18081/` again and type queries like 哎哟, you should see the newly added document.

### CODE REVIEW

**Comment by ghost** [comment]
I think the easiest way to start reviewing this PR is from index.go. :-)

**Comment by chongw** [comment]
really dump q: not able to run test even after go get several different packages.  (adamzy, davecgh, huichen, mattn, olekukonko, pmezard, stretchr)

```
running go test -v gives 
./index_test.go:93: s.LoadDictionary(path.Join(gosrc(), "github.com/huichen/sego/data/dictionary.txt")) used as value
FAIL    github.com/wangkuiyi/weakand [build failed]
```

**Update from Yi**

I guess this has been solved? :) Sorry that it is because that my PR to `github.com/huichen/sego` has not yet been merged to its master branch. So we will have to manually checkout my branch `fs` in repo `github.com/wangkuiyi/sego`

File: index.go [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+func tokenize(doc string, sgmt *sego.Segmenter) []string {
+	var terms []string
+	for _, seg := range sgmt.Segment([]byte(doc)) {
+		term := seg.Token().Text()
+		if !AllPunctOrSpace(term) {
+			terms = append(terms, term)
+		}
+	}
+	return terms
+}
```
Comment (chongw) [inline index.go:150]: just in general, do we have a general concept of meta data here? for example, time stamp etc.

**Comment by chqsark** [comment]
是不是把index，search，等都放到各自的目录底下？现在有点乱

**Update from Yi**
Currently, all source files in the root directory of this repo use data types and functions defined in each other. I do not yet have a good idea how to decouple the dependencies. Usually, when dependencies are hard to decouple, which mean tightly coupled, we put these files in a single Go package.

File: index.go (hunk 1 above)
Comment (wangkuiyi) [inline index.go:150]: Not yet. I agree that we will have more meta data in the future. Currently this PR implements the most basic functionalities only. I guess we will have a lot of improvements to do after this PR.

**Comment by chqsark** [comment]
LGTM

**Comment by chongw** [comment]
LGTM. 终于看完了，同时学了不少GO。

**Comment by chongw** [comment]
travis-ci  是个专门做continous integration的server么?

**Comment by wangkuiyi** [comment]
@chongw 是的。Travis CI是Github可以整合的一个做continuous integration的web 服务。

**Pull request merged**